src/x86/intel_joule_expansion.c

```diff
@@ -145,6 +181,115 @@ mraa_joule_expansion_board()
     b->spi_bus[2].bus_id = 32766;
     b->spi_bus[2].slave_s = 1;
 
+    // Aio shield detection
+    // Only if i2c_designware.0 was successfully found
+    if (i2c_aio_bus != -1) {
+
+        syslog(LOG_NOTICE, "Attempting shield autodetection on I2C bus %d...", i2c_aio_bus);
+
+        i2c_smbus_ioctl_data_t sm;
+        i2c_smbus_data_t d;
+        int fp;
+
+        char i2c_fh[64];
+        snprintf(i2c_fh, 64, "/dev/i2c-%d", i2c_aio_bus);
+
+        sm.read_write = I2C_SMBUS_READ;
+        sm.command = 0x01;
+        sm.size = I2C_SMBUS_WORD_DATA; // LSB first
+        sm.data = &d;
+
+        if ((fp = open(i2c_fh, O_RDWR)) < 1)
+            syslog(LOG_ERR, "Failed to open I2C bus: %s", strerror(errno));
+        if (ioctl(fp, I2C_SLAVE_FORCE, 0x49) < 0)
+            syslog(LOG_ERR, "Failed to set I2C address: %s", strerror(errno));
+        if (ioctl(fp, I2C_SMBUS, &sm) < 0) {
+            syslog(LOG_NOTICE, "No device at I2C address 0x49");
+
+            // no DFRobot shield, try Grove
+            if (ioctl(fp, I2C_SLAVE_FORCE, 0x48) < 0)
+                syslog(LOG_ERR, "Failed to set I2C address: %s", strerror(errno));
+            if (ioctl(fp, I2C_SMBUS, &sm) < 0)
+                syslog(LOG_NOTICE, "No device at I2C address 0x48");
+                // no shield
+            else {
+                // load ads1015 at 0x48
+                syslog(LOG_NOTICE, "Loading ti-ads1015 module for Grove Shield");
+                if (system("modprobe ti-ads1015") < 0) {
```

Comment:
This is the tricky part. Modprobe wouldn't be needed if the ads1015 wasn't double aliased (hwmon/ads1015 vs. iio/ti-ads1015). I tend to disagree with using linux/module.h as we'll just end up rewriting what insmod and modprobe already do. The system call with a hardcoded string I'm not worried about, the rest of them on the other hand I would happily replace with a i2c_new_device() + MRAA's IIO functions if all needed functionality is there.